src/PowerShellEditorServices/Session/Host/TerminalPSHostRawUserInterface.cs

```diff
@@ -184,11 +145,87 @@ public override Size MaxWindowSize
         /// <returns>A KeyInfo struct with details about the current keypress.</returns>
         public override KeyInfo ReadKey(ReadKeyOptions options)
         {
-            Logger.Write(
-                LogLevel.Warning,
-                "PSHostRawUserInterface.ReadKey was called");
+            // Converts ConsoleKeyInfo objects to KeyInfo objects and caches key down events
+            // for the next key up request.
+            KeyInfo ProcessKey(ConsoleKeyInfo key, bool isDown)
+            {
+                // Translate ConsoleModifiers to ControlKeyStates
+                ControlKeyStates states = default;
+                if ((key.Modifiers & ConsoleModifiers.Alt) != 0)
+                {
+                    states |= ControlKeyStates.LeftAltPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Control) != 0)
+                {
+                    states |= ControlKeyStates.LeftCtrlPressed;
+                }
+
+                if ((key.Modifiers & ConsoleModifiers.Shift) != 0)
+                {
+                    states |= ControlKeyStates.ShiftPressed;
+                }
 
-            throw new System.NotImplementedException();
+                var result = new KeyInfo((int)key.Key, key.KeyChar, states, isDown);
+                if (isDown)
+                {
+                    this.lastKeyDown = result;
```

Comment:
The returned value from this local function is returned immediately at all call sites of this local function, which means the cached KeyDown returned below in `if (includeUp && this.lastKeyDown != null)` has already been returned previously.

Reply:
@daxian-dbw 

> The returned value from this local function is returned immediately at all call sites of this local function, which means the cached KeyDown returned below in if (includeUp && this.lastKeyDown != null) has already been returned previously.

Yeah, that's an attempt match the behavior of `ConsoleHost` without dipping into native API's.

Here's an example to demonstrate what I'm trying emulate.

```powershell
while ($true) {
    $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyDown')
    $Host.UI.RawUI.ReadKey('NoEcho, IncludeKeyUp')
}
```

With that running outside of PSES, every one key press will return at least two `KeyInfo` objects.  One for down, and one for up. Example output:

```raw
VirtualKeyCode Character ControlKeyState KeyDown
-------------- --------- --------------- -------
            70         f       NumLockOn    True
            70         f       NumLockOn   False
```

Now obviously `ConsoleHost` does this a lot better using native API's directly, but this saves a lot of complexity.  The only difference in behavior is when holding down a key with both `IncludeKeyUp` and `IncludeKeyDown` set.  In that scenario `ConsoleHost` will continuously return key down `KeyInfo` objects until they key is released where we will continuously return both key down *and* key up objects.  That difference is negated a little bit though. Even `ConsoleHost` act the same under `winpty` as it doesn't translate input one to one.

Reply:
Yeah on non-consolehost I think it typically only supports KeyDown events and not KeyUp at all.

If I tried that while loop on macOS, I only get KeyDown events.

Reply:
Just to clarify because the terms are confusing, when I say`ConsoleHost` I'm referring to `Microsoft.PowerShell.ConsoleHost` (the `PSHost` started by `powershell.exe`/`pwsh`) not `conhost.exe`

Reply:
ah sorry. Yeah I meant non-conhost sorry.

Reply:
Gotcha! Thanks for the clarification!